File: apps/android/feature/analytics/data/src/main/java/com/scribblefit/feature/analytics/data/repository/AnalysisRepositoryImpl.kt
package com.scribblefit.feature.analytics.data.repository

import com.scribblefit.core.database.dao.InsightsCacheDao
import com.scribblefit.core.database.entity.InsightsCacheEntity
import com.scribblefit.feature.ai.domain.model.AnalysisSuggestion
import com.scribblefit.feature.ai.domain.model.AnalysisSummary
import com.scribblefit.feature.ai.domain.model.ExerciseInsight
import com.scribblefit.feature.ai.domain.model.SummaryPeriod
import com.scribblefit.feature.analytics.domain.repository.AnalysisRepository
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.map
import kotlinx.serialization.json.Json
import javax.inject.Inject
import javax.inject.Singleton

private const val KEY_HOME_SUGGESTION = "home_suggestion"

@Singleton
class AnalysisRepositoryImpl @Inject constructor(
    private val dao: InsightsCacheDao,
    private val json: Json
) : AnalysisRepository {

    override fun getHomeSuggestion(): Flow<AnalysisSuggestion?> =
        dao.getByKey(KEY_HOME_SUGGESTION).map { entity ->
            entity?.jsonData?.let { runCatching { json.decodeFromString<AnalysisSuggestion>(it) }.getOrNull() }
        }

    override fun getSummary(period: SummaryPeriod): Flow<AnalysisSummary?> =
        dao.getByKey("summary_${period.name.lowercase()}").map { entity ->
            entity?.jsonData?.let { runCatching { json.decodeFromString<AnalysisSummary>(it) }.getOrNull() }
        }

    override fun getExerciseInsight(exerciseId: String): Flow<ExerciseInsight?> =
        dao.getByKey("exercise_insight_$exerciseId").map { entity ->
            entity?.jsonData?.let { runCatching { json.decodeFromString<ExerciseInsight>(it) }.getOrNull() }
        }

    override suspend fun saveHomeSuggestion(suggestion: AnalysisSuggestion) {
        dao.upsert(InsightsCacheEntity(KEY_HOME_SUGGESTION, json.encodeToString(suggestion), System.currentTimeMillis()))
    }

    override suspend fun saveSummary(summary: AnalysisSummary) {
        val key = "summary_${summary.period.name.lowercase()}"
        dao.upsert(InsightsCacheEntity(key, json.encodeToString(summary), System.currentTimeMillis()))
    }

    override suspend fun saveExerciseInsight(insight: ExerciseInsight) {
        val key = "exercise_insight_${insight.exerciseId}"
        dao.upsert(InsightsCacheEntity(key, json.encodeToString(insight), System.currentTimeMillis()))
    }

    override suspend fun clearOldInsights() {
        dao.deleteAll()
    }
}

File: apps/ios/ScribbleFit/Sources/Features/Analytics/Data/AnalysisRepositoryImpl.swift
import Combine
import Foundation

@MainActor
public final class AnalysisRepositoryImpl: AnalysisRepository {
    private let database: ScribbleFitDatabase
    nonisolated(unsafe) private let homeSuggestionSubject = CurrentValueSubject<AnalysisSuggestion?, Never>(nil)
    private let decoder = JSONDecoder()
    private let encoder = JSONEncoder()

    public init(database: ScribbleFitDatabase) {
        self.database = database
        if let cached = database.getInsightByKey(key: "home_suggestion"),
           let data = cached.jsonData.data(using: .utf8),
           let suggestion = try? decoder.decode(AnalysisSuggestion.self, from: data) {
            homeSuggestionSubject.send(suggestion)
        }
    }

    public nonisolated func getHomeSuggestion() -> AnyPublisher<AnalysisSuggestion?, Never> {
        homeSuggestionSubject.eraseToAnyPublisher()
    }

    public func getSummary(period: SummaryPeriod) async throws -> AnalysisSummary? {
        guard let cache = database.getInsightByKey(key: "summary_\(period.rawValue)"),
              let data = cache.jsonData.data(using: .utf8) else { return nil }
        return try? decoder.decode(AnalysisSummary.self, from: data)
    }

    public func getExerciseInsight(exerciseId: String) async throws -> ExerciseInsight? {
        guard let cache = database.getInsightByKey(key: "exercise_insight_\(exerciseId)"),
              let data = cache.jsonData.data(using: .utf8) else { return nil }
        return try? decoder.decode(ExerciseInsight.self, from: data)
    }

    public func saveHomeSuggestion(_ suggestion: AnalysisSuggestion) async throws {
        let data = try encoder.encode(suggestion)
        guard let json = String(data: data, encoding: .utf8) else { return }
        database.upsertInsight(InsightsCache(key: "home_suggestion", jsonData: json))
        homeSuggestionSubject.send(suggestion)
    }

    public func saveSummary(_ summary: AnalysisSummary) async throws {
        let data = try encoder.encode(summary)
        guard let json = String(data: data, encoding: .utf8) else { return }
        database.upsertInsight(InsightsCache(key: "summary_\(summary.period.rawValue)", jsonData: json))
    }

    public func saveExerciseInsight(_ insight: ExerciseInsight) async throws {
        let data = try encoder.encode(insight)
        guard let json = String(data: data, encoding: .utf8) else { return }
        database.upsertInsight(InsightsCache(key: "exercise_insight_\(insight.exerciseId)", jsonData: json))
    }

    public func clearOldInsights() async throws {
        database.clearInsights()
        homeSuggestionSubject.send(nil)
    }
}
